lib/std/array_list.zig

```diff
@@ -153,131 +145,143 @@ pub fn AlignedArrayList(comptime T: type, comptime alignment: ?u29) type {
             };
+            self.capacity = new_len;
         }
 
         pub fn ensureCapacity(self: *Self, new_capacity: usize) !void {
-            var better_capacity = self.capacity();
+            var better_capacity = self.capacity;
             if (better_capacity >= new_capacity) return;
+
             while (true) {
                 better_capacity += better_capacity / 2 + 8;
                 if (better_capacity >= new_capacity) break;
             }
-            self.items = try self.allocator.realloc(self.items, better_capacity);
+            
+            const new_memory = try self.allocator.realloc(self.memory(), better_capacity);
+            self.items.ptr = new_memory.ptr;
+            self.capacity = new_memory.len;
         }
 
         /// Increases the array's length to match the full capacity that is already allocated.
         /// The new elements have `undefined` values. This operation does not invalidate any
         /// element pointers.
         pub fn expandToCapacity(self: *Self) void {
-            self.len = self.items.len;
+            self.items.len = self.capacity;
         }
 
         /// Increase length by 1, returning pointer to the new item.
         /// The returned pointer becomes invalid when the list is resized.
         pub fn addOne(self: *Self) !*T {
-            const new_length = self.len + 1;
-            try self.ensureCapacity(new_length);
+            const newlen = self.items.len + 1;
+            try self.ensureCapacity(newlen);
             return self.addOneAssumeCapacity();
         }
 
         /// Increase length by 1, returning pointer to the new item.
         /// Asserts that there is already space for the new item without allocating more.
         /// The returned pointer becomes invalid when the list is resized.
         pub fn addOneAssumeCapacity(self: *Self) *T {
-            assert(self.len < self.capacity());
-            const result = &self.items[self.len];
-            self.len += 1;
-            return result;
+            assert(self.items.len < self.capacity);
+
+            self.items.len += 1;
+            return &self.items[self.items.len - 1];
         }
 
         /// Remove and return the last element from the list.
         /// Asserts the list has at least one item.
         pub fn pop(self: *Self) T {
-            self.len -= 1;
-            return self.items[self.len];
+            const val = self.items[self.items.len - 1];
+            self.items.len -= 1;
+            return val;
         }
 
         /// Remove and return the last element from the list.
         /// If the list is empty, returns `null`.
         pub fn popOrNull(self: *Self) ?T {
-            if (self.len == 0) return null;
+            if (self.items.len == 0) return null;
             return self.pop();
         }
+        
+        fn memory(self: Self) Slice {
+            return self.items.ptr[0..self.capacity];
```

Comment:
Might be worth a comment on why this procedure is necessary; `items` only refers to the elements that have been appended, and not the entire capacity, hence the "unsafe" slicing.

In the same vein, maybe calling it `allocatedRegion`, or something that highlights this fact more easily, might be clearer?